Sources/Runtime/arc.cpp

```diff
@@ -0,0 +1,144 @@
+//
+//  arc.cpp
+//  trill
+//
+//  Created by Harlan Haskins on 3/6/17.
+//  Copyright © 2017 Harlan. All rights reserved.
+//
+
+#include "arc.h"
+#include "trill.h"
+#include <cstdlib>
+#include <atomic>
+#include <mutex>
+
+using namespace trill;
+
+/**
+ A RAII class that will lock a mutex when it's constructed
+ and unlock the mutex once it's destroyed.
+ */
+struct LockRAII {
+  pthread_mutex_t mutex;
+
+  /**
+   Locks the provided mutex
+   */
+  LockRAII(pthread_mutex_t mutex): mutex(mutex) {
+    pthread_mutex_lock(&mutex);
+  }
+
+  /**
+   Unlocks the provided mutex
+   */
+  ~LockRAII() {
+    pthread_mutex_unlock(&mutex);
+  }
+};
+
+/**
+ A RefCountBox contains
+  - A retain count
+  - A mutex used to synchronize retains and releases
+  - A variably-sized payload that is not represented by a data member.
+
+  It is meant as a hidden store for retain count data alongside the allocated
+  contents of an indirect type.
+
+  Trill will always see this as:
+      [retainCount][mutex][payload]
+                          ^~ indirect type "begins" here
+ */
+struct RefCountBox {
+  uint32_t retainCount;
+  pthread_mutex_t mutex;
+
+  /// Finds the payload by walking to the end of the data members of this box.
+  void *value() {
+    return reinterpret_cast<void *>(reinterpret_cast<uintptr_t>(this) +
+                                    sizeof(RefCountBox));
+  }
+};
+
+/**
+ A convenience struct that performs the arithmetic necessary to work with a
+ \c RefCountBox.
+ */
+struct RefCounted {
```

Comment:
Definitely would advise looking at what the ObjC RT does here -- I trust Greg way more than either of us :P